include/alpaka/mem/view/Accessor.hpp

```diff
@@ -0,0 +1,186 @@
+        struct MpContains<List<Head, Tail...>, Value>
+        {
+            static constexpr bool value = std::is_same<Head, Value>::value || MpContains<List<Tail...>, Value>::value;
+        };
+    } // namespace internal
+
+    //! Constrains an existing accessor with multiple access modes to the specified access modes.
+    // TODO: currently only allows constraining down to 1 access mode
+    template<
+        typename TNewAccessMode,
+        typename TMemoryHandle,
+        typename TElem,
+        typename TBufferIdx,
+        std::size_t TDim,
+        typename... TPrevAccessModes>
+    ALPAKA_FN_HOST_ACC auto accessWith(
+        const Accessor<TMemoryHandle, TElem, TBufferIdx, TDim, std::tuple<TPrevAccessModes...>>& acc)
+    {
+        static_assert(
+            internal::MpContains<std::tuple<TPrevAccessModes...>, TNewAccessMode>::value,
+            "The accessed accessor must already contain the requested access mode");
+        return Accessor<TMemoryHandle, TElem, TBufferIdx, TDim, TNewAccessMode>{acc};
+    }
+
+    //! Constrains an existing accessor to the specified access modes.
+    // constraining accessor to the same access mode again just passes through
+    template<typename TNewAccessMode, typename TMemoryHandle, typename TElem, typename TBufferIdx, std::size_t TDim>
+    ALPAKA_FN_HOST_ACC auto accessWith(const Accessor<TMemoryHandle, TElem, TBufferIdx, TDim, TNewAccessMode>& acc)
+    {
+        return acc;
+    }
+
+    //! Creates a read-write accessor for the given memory object (view, buffer, ...) or accessor.
+    template<typename TMemoryObjectOrAccessor>
+    ALPAKA_FN_HOST_ACC auto access(TMemoryObjectOrAccessor&& viewOrAccessor)
+    {
+        return accessWith<ReadWriteAccess>(std::forward<TMemoryObjectOrAccessor>(viewOrAccessor));
+    }
+
+    //! Creates a read-only accessor for the given memory object (view, buffer, ...) or accessor.
+    template<typename TMemoryObjectOrAccessor>
+    ALPAKA_FN_HOST_ACC auto readAccess(TMemoryObjectOrAccessor&& viewOrAccessor)
+    {
+        return accessWith<ReadAccess>(std::forward<TMemoryObjectOrAccessor>(viewOrAccessor));
+    }
+
+    //! Creates a write-only accessor for the given memory object (view, buffer, ...) or accessor.
+    template<typename TMemoryObjectOrAccessor>
+    ALPAKA_FN_HOST_ACC auto writeAccess(TMemoryObjectOrAccessor&& viewOrAccessor)
+    {
+        return accessWith<WriteAccess>(std::forward<TMemoryObjectOrAccessor>(viewOrAccessor));
+    }
+
+    //! An alias for an accessor accessing a buffer on the given accelerator.
+    template<
+        typename TAcc,
+        typename TElem,
+        std::size_t TDim,
+        typename TAccessModes = ReadWriteAccess,
+        typename TIdx = Idx<TAcc>>
+    using BufferAccessor = Accessor<
+        MemoryHandle<decltype(
+            accessWith<TAccessModes>(alpaka::core::declval<Buf<TAcc, TElem, DimInt<TDim>, TIdx>>()))>,
+        TElem,
+        TIdx,
+        TDim,
+        TAccessModes>;
```

Comment:
One of the remaining issues is here. `BufferAccessor` depends on the type returned from `accessWith` via `decltype`. `BufferAccessor` typically occurs at the kernel interface and thus in `__device__` code. However, `accessWith` is `__host__ __device__` and may call into `__host__` functions (like `getPtrNative`), e.g. for accessing a buffer object (see: e.g. https://github.com/bernhardmgruber/alpaka/runs/2389196556?check_suite_focus=true#step:5:4364). clang < 9 rejects this and fails to compile. Nvcc and clang >= 9 accept this in all versions.

We have a couple of options to resolve this:
1. make the offending functions `__host__ __device` for clang < 9. For now I see `getPtrNative` and `getExtent` to require this annotation, but there might be further functions called from there.
2. avoid the `decltype` and make `BufferAccessor` use a trait that needs to be implemented for each accessor. This is more work on the implementer and duplicates information between `accessWith` and this new trait.
3. stop supporting clang < 9 in CUDA mode.

Since this seems to be a clang bug and I would not want to complicate the overall implementation because of that, I would recommend either 1. or 3.

@psychocoderHPC I would especially request your input here please :)

Reply:
I would not say it is a clang bug. Clang always said they pars things different compared to nvcc. 
https://llvm.org/docs/CompileCudaWithLLVM.html#overloading-based-on-host-and-device-attributes

I do not see an issue with the Way 2. If we can not derive enough knowledge from the language we need to get it from somewhere.

Way 3: Should be chosen wisely to avoid that we jump always to a new compiler if we observe an issue. 
Never the less clang-cuda 8 is only supporting <=CUDA 10.0 therefore dropping support should not be a real issue.
If you like to remove clang-cuda <=8 please open a new PR for it.

Reply:
> If you like to remove clang-cuda <=8 please open a new PR for it.

Ok, let's go: #1300

Reply:
Resolved because #1300 was merged and this PR rebased on it.